Update mysql and snowflake tutorials for st.connection

sfc-gh-jcarroll opened the pull request:
# TODOS

- [x] The MySQL tutorial in particular needs an older copy to be archived to link in the blog post for comparison. Not sure how best to do that (i can easily make a new page and copy the old page with a different slug but want to confirm with @snehankekre if that approach makes sense, or something else). In theory could also use the wayback machine.
- [x] I didn't test the e2e setup although I did test the final code locally. I can do this but didn't know if @snehankekre wanted to do it personally.

## 📚 Context

Updating database tutorials to use st.connection. Depends on pages in https://github.com/streamlit/docs/pull/622

## 🧠 Description of Changes

* Update the MySQL tutorial to use st.connection API
* Update the Snowflake tutorial to use st.connection API

**Revised:**

Here's the main update, there are some other updates too

![image](https://user-images.githubusercontent.com/116604821/234416576-295cb2ae-4103-4040-a245-7859fb8f117e.png)

**Current:**

https://docs.streamlit.io/knowledge-base/tutorials/databases/mysql

## 💥 Impact

Changes the tutorial setup and code

Size:

- [ ] Small <!-- Small bug fix or small edit to existing code that amounts to few lines) -->
- [x] Not small <!-- Everything else -->

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-jcarroll commented on content/kb/tutorials/databases/mysql.md:
```diff
@@ -79,32 +81,19 @@ Copy the code below to your Streamlit app and run it. Make sure to adapt `query`
 # Print results.
-for row in rows:
-    st.write(f"{row[0]} has a :{row[1]}:")
+for person, pet in zip(df['person'], df['pet']):
+    st.write(f"{person} has a :{pet}:")
```
Pandas experts in the house - should I do this approach or something else?

snehankekre replied:
Used `df.itertuples()` a85f85a and 798e125.

Btw, in the Snowflake app code here `st.write(f"{row.person} has a :{row.pet}:")`, should PERSON and PET be capitalized? Need to verify this but IIRC Snowflake returns capitalized column names?

sfc-gh-jcarroll replied:
snowpark-python DOES return capitals - fixed. Good catch

snowflake-sqlalchemy + SQLConnection.query() returns lower case, so I left it that way for this example.

snehankekre commented on content/kb/tutorials/databases/mysql.md:
````diff
@@ -79,32 +81,19 @@ Copy the code below to your Streamlit app and run it. Make sure to adapt `query`
 ```
 
-See `st.cache_data` above? Without it, Streamlit would run the query every time the app reruns (e.g. on a widget interaction). With `st.cache_data`, it only runs when the query changes or after 10 minutes (that's what `ttl` is for). Watch out: If your database updates more frequently, you should adapt `ttl` or remove caching so viewers always see the latest data. Learn more in [Caching](/library/ ...
+See `st.experimental_connection` above? This handles secrets retrieval, setup, query caching and retries. In this case, we set `ttl=600` to ensure the query result is cached for no longer than 10 minutes. Watch out: If your database updates more frequently, you should adapt `ttl` or remove caching so viewers always see the latest data. Learn more in[Connecting to data](/library/advanced-features/ ...
````
```suggestion
See `st.experimental_connection` above? This handles secrets retrieval, setup, query caching and retries. In this case, we set `ttl=600` to ensure the query result is cached for no longer than 10 minutes. Watch out: If your database updates more frequently, you should adapt `ttl` or remove caching so viewers always see the latest data. Learn more in [Connecting to data](/library/advanced-features/connecting-to-data) and [Caching](/library/advanced-features/caching).
```

Adds a space between `in` and `Connection to data`

snehankekre commented:
Kudos for the update @sfc-gh-jcarroll!! Did a cursory review and the changes LGTM :)

> I can do this but didn't know if @snehankekre wanted to do it personally.

Would be great if you are able to. Else, I do it early next week too. 

> The MySQL tutorial in particular needs an older copy to be archived to link in the blog post for comparison. 

We could link to an internet archived url if the purpose is comparison in the blog post. Or link to an older netlify deploy preview.  E.g. https://deploy-preview-620--streamlit-docs.netlify.app/knowledge-base/tutorials/databases/mysql

The number after `deploy-preview` in the url corresponds to the PR number. You could decrement 620 to much older versions as needed.

sfc-gh-jcarroll commented:
Internet archive for the win! I think we can just use this: https://web.archive.org/web/20230330050343/https://docs.streamlit.io/knowledge-base/tutorials/databases/mysql#write-your-streamlit-app

I like that better than netlify preview - easier to understand and potentially more durable.

sfc-gh-jcarroll commented:
✅ MySQL confirmed locally, found and fixed one bug on a column name mismatch

sfc-gh-jcarroll commented:
✅ Snowflake examples tested locally

I also added a section/example on using the Snowpark Session, and split out the snowflake SQL secrets setup to be a bit clearer.

Should be mergeable from my perspective

vdonato commented on content/kb/tutorials/databases/mysql.md:
````diff
@@ -79,32 +81,19 @@ Copy the code below to your Streamlit app and run it. Make sure to adapt `query`
 ```
 
-See `st.cache_data` above? Without it, Streamlit would run the query every time the app reruns (e.g. on a widget interaction). With `st.cache_data`, it only runs when the query changes or after 10 minutes (that's what `ttl` is for). Watch out: If your database updates more frequently, you should adapt `ttl` or remove caching so viewers always see the latest data. Learn more in [Caching](/library/ ...
+See `st.experimental_connection` above? This handles secrets retrieval, setup, query caching and retries. In this case, we set `ttl=600` to ensure the query result is cached for no longer than 10 minutes. Watch out: If your database updates more frequently, you should adapt `ttl` or remove caching so viewers always see the latest data. Learn more in [Connecting to data](/library/advanced-features ...
````
Hm, it might be useful to call out removing caching is done in this case by setting `ttl=None`. Otherwise, I could see some users just deleting the kwarg and having results saved indefinitely.

vdonato commented on content/kb/tutorials/databases/snowflake.md:
```diff
@@ -60,107 +52,157 @@ Be sure to highlight or select **all** the queries (lines 1-10) before clicking
-## Write your Streamlit app
+import streamlit as st
+
+# Initialize connection and access underlying Snowpark session.
+session = st.experimental_connection('snowpark').session
```
Hm, I think we probably want to direct people toward using `safe_session` here since I'd imagine that many users following this tutorial won't always know what assumptions they're able to make about thread safety / be expected to be able to guarantee access to the `session` object is thread safe.

vdonato commented on content/kb/tutorials/databases/snowflake.md:
````diff
@@ -60,107 +52,157 @@ Be sure to highlight or select **all** the queries (lines 1-10) before clicking
+```
 
-5. On Streamlit Community Cloud, select the appropriate version of Python for your app by clicking "Advanced settings" before you deploy the app:
+See `st.experimental_connection` above? This handles secrets retrieval, setup, query caching and retries. In this case, we set `ttl=600` to ensure the query result is cached for no longer than 10 minutes. Watch out: If your database updates more frequently, you should adapt `ttl` or remove caching so viewers always see the latest data. Learn more in [Connecting to data](/library/advanced-features ...
````
Same comment about `ttl=None` here

vdonato reviewed (approved):
A few small comments, but otherwise LGTM

sfc-gh-jcarroll commented:
@vdonato updates look alright to address your feedback? Thanks!

vdonato reviewed (approved):
👍 LGTM
